### PR TITLE
Get real Flash Chip Size via API call

### LINKED ISSUE #7157: ESP32 Flash size read via API is wrong when flash is in modes `qio, dio, qout`
### Board

esp32 devkit 4 MB

### Device Description

Plain Devkit board, nothing connected

### Hardware Configuration

none

### Version

v2.0.4

### IDE Name

Platformio latest release

### Operating System

macOS 12.5.1

### Flash frequency

40 Mhz

### PSRAM enabled

no

### Upload speed

115200

### Description

Get flash size correctly when in flash modes `qio, qout, dio` as it does when in flash mode `dout`.
The standard Arduino function get flash size is useless in most cases, since it just reads the values which are set from the IDE and it does not detect the real flash size

### Sketch

Sketch to test https://github.com/Jason2866/ESP32_Show_Info
Can be run with Arduino IDE too.
The test sketch shows too that flash mode is not correct when queried via the API (qio and qout never shown)

### Debug Message

```plain
First (correct) output in mode dout 

entry 0x400805c4
can't allocate memory with malloc

ESP32 SDK: v4.4.3
ESP32 CPU FREQ: 240 MHz
ESP32 APB FREQ: 80.0 MHz
ESP32 FLASH CHIP ID: 1458392
ESP32 FLASH REAL SIZE: 4.00 MB
ESP32 FLASH SIZE (MAGIC BYTE): 4.00 MB
ESP32 FLASH MODE: 3, 0=QIO, 1=QOUT, 2=DIO, 3=DOUT
ESP32 RAM SIZE: 366.19 KB
ESP32 FREE RAM: 342.56 KB
ESP32 MAX RAM ALLOC: 119.99 KB
ESP32 FREE PSRAM: 0.00 KB
```
Second (wrong) output in mode qio (shown wrong dio too)
```
entry 0x400805d0
can't allocate memory with malloc

ESP32 SDK: v4.4.3
ESP32 CPU FREQ: 240 MHz
ESP32 APB FREQ: 80.0 MHz
ESP32 FLASH CHIP ID: 729324
ESP32 FLASH REAL SIZE: 0.00 MB
ESP32 FLASH SIZE (MAGIC BYTE): 4.00 MB
ESP32 FLASH MODE: 2, 0=QIO, 1=QOUT, 2=DIO, 3=DOUT
ESP32 RAM SIZE: 366.19 KB
ESP32 FREE RAM: 342.56 KB
ESP32 MAX RAM ALLOC: 119.99 KB
ESP32 FREE PSRAM: 0.00 KB
```
```


### Other Steps to Reproduce

Running the example sketch

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

### PR DESCRIPTION
**Pull request by Jason2866**
actually only the value from the IDE which has been set is read from magic header.
Now the real size is read from the flash chip.

Credits got to @Staars

Fixes #7157

### Checklist
1. [x] Please provide specific title of the PR describing the change, including the component name (*eg. „Update of Documentation link on Readme.md“*)
2. [x] Please provide related links (*eg. Issue which will be closed by this Pull Request*)
3. [x] Please **update relevant Documentation** if applicable
4. [x] Please check [Contributing guide](https://docs.espressif.com/projects/arduino-esp32/en/latest/contributing.html)

### CODE REVIEW

**Review by me-no-dev — APPROVED** [review]

**Comment by me-no-dev** [comment]
Have you tried this on OPI flash?

**Comment by Jason2866** [comment]
> Have you tried this on OPI flash?

No, sorry to say i have no S3 with OPI

File: cores/esp32/Esp.cpp [hunk 1]
```diff
@@ -305,13 +305,17 @@ const char * EspClass::getSdkVersion(void)
...
+
 uint32_t EspClass::getFlashChipSize(void)
 {
-    esp_image_header_t fhdr;
-    if(flashRead(ESP_FLASH_IMAGE_BASE, (uint32_t*)&fhdr, sizeof(esp_image_header_t)) && fhdr.magic != ESP_IMAGE_HEADER_MAGIC) {
-        return 0;
-    }
-    return magicFlashChipSize(fhdr.spi_size);
+  uint32_t id = (ESP_getFlashChipId() >> 16) & 0xFF;
+  return 2 << (id - 1);
```
Comment (SuGlider) [inline cores/esp32/Esp.cpp:318]: Thinking about the question that @me-no-dev made about OPI...

Maybe if the call to `ESP_getFlashChipId()` returns some invalid information regarding the Flash Size, this API could "fall back" to the `EmagicFlashChipSize()` way.

Any opinion on it?

Comment (Jason2866) [reply cores/esp32/Esp.cpp:318]: Good idea. It needs testing if it fails with OPI or not. You have at least one working one :-)

Comment (Jason2866) [reply cores/esp32/Esp.cpp:318]: Magic byte will deliver a wrong value for OPI anyways.

**Comment by Jason2866** [comment]
Just my 2 cents. Real S3 support is missing in mandantory tools https://github.com/espressif/esptool/issues/730
https://github.com/espressif/esptool/issues/768

**Comment by VojtechBartoska** [comment]
@P-R-O-C-H-Y will test with OPI

**Comment by P-R-O-C-H-Y** [comment]
Runned @Jason2866 's sketch from linked issue on ESP32S3 with 32MB OPI flash. In Arduino IDE tools menu a have selected 16 MB Flash size (no 32 MB option). So seems to be working great with OPI flash :) Good job @Jason2866 👍 

```
ESP32 SDK: v4.4.2-1-gce1fde4b3b
ESP32 DEVICE: ESP32-S3
ESP32 CPU FREQ: 240 MHz
ESP32 XTAL FREQ: 40 MHz
ESP32 APB FREQ: 80.0 MHz
ESP32 FLASH CHIP ID: 3768514
ESP32 FLASH CHIP FREQ: 80.0 MHz
ESP32 FLASH REAL SIZE: 32.00 MB
ESP32 FLASH SIZE (MAGIC BYTE): 16.00 MB
ESP32 FLASH REAL MODE: Fast
ESP32 FLASH MODE (MAGIC BYTE): 3, 0=QIO, 1=QOUT, 2=DIO, 3=DOUT
ESP32 RAM SIZE: 388.85 KB
ESP32 FREE RAM: 364.94 KB
ESP32 MAX RAM ALLOC: 295.99 KB
ESP32 FREE PSRAM: 7139.33 KB
```

**Review by P-R-O-C-H-Y — APPROVED** [review]

**Review by SuGlider — APPROVED** [review]

**Pull request merged**